src/System.Linq/src/System/Linq/Distinct.cs

```diff
@@ -116,6 +116,11 @@ public List<TSource> ToList()
                 return FillSet().ToList();
             }
 
+            public HashSet<TSource> ToHashSet(IEqualityComparer<TSource> comparer)
+            {
+                return FillSet().ToHashSet(comparer);
```

Comment:
Yes, that's the way to do the comparison on comparers.

On the use of `HashSet<T>` instead of the internal `Set<T>`, AFAICT the only reason `Set<T>` exists is linq and `HashSet<T>` were implemented independently at the same time and so the former couldn't make use of the latter. In Expressions we changed to just use HashSet and here we optimised `Set<T>` for the uses made of it at the cost of not making it work in proper use-cases that don't come up. `HashSet<T>` does have some slight extra costs (it tracks changes to throw in an enumeration it is changed within) so it would have a performance impact, but not a severe one. I don't have a strong opinion on the disadvantage of that vs the advantage of being lighter and using the established collection for such purposes.